api/app/Extensions/Revisionable/ChangeloggableTrait.php

```diff
@@ -0,0 +1,282 @@
+            foreach ($childModels as $model) {
+                array_push(
+                    $revisions,
+                    $this->prepareRevision($key, null, $model->toJson())
+                );
+            }
+
+            if (count($revisions)) {
+                $this->cleanupRevisions(count($revisions));
+
+                $this->dbInsert($revisions);
+            }
+        }
+    }
+
+    /**
+     * Invoked before a model is synced.
+     *
+     * @param  string $key
+     *
+     * @return void
+     */
+    public function preSync($relation)
+    {
+        if ($this->isRevisionEnabled()
+            && $this->isRelationRevisionable($relation)
+        ) {
+            // Get only the IDs from the relationship
+            $ids = array_keys($this->$relation->modelKeys());
+
+            // And store them under the relationship name
+            $this->originalData = [$relation => $ids];
+        }
+    }
+
+    /**
+     * Called after a model is successfully synced.
+     *
+     * @param  string $key
+     * @param  array  $ids
+     *
+     * @return void
+     */
+    public function postSync($key, array $ids)
+    {
+        if (($this->isRevisionEnabled())
+            && (!$this->isLimitReached() || $this->isRevisionCleanup())
+            && array_key_exists($key, $this->originalData)
+        ) {
+            $revision = $this->prepareRevision($key, json_encode(array_get($this->originalData, $key)), json_encode($ids));
+
+            $this->cleanupRevisions();
+
+            $this->dbInsert($revision);
+        }
+    }
+
+    /**
+     * Called after a child model is deleted.
+     *
+     * @param  string    $key
+     * @param  BaseModel $model
+     *
+     * @return void
+     */
+    public function postDeleteOneChild($key, $model)
+    {
+        if ($this->isRevisionEnabled()) {
+            $revision = $this->prepareRevision($key, $model, null);
+
+            $this->cleanupRevisions();
+            $this->dbInsert($revision);
+        }
+    }
+
+    /**
+     * Determines if the relationship is revisionable.
+     *
+     * @param  string $relation
+     *
+     * @return boolean
+     */
+    protected function isRelationRevisionable($relation)
+    {
+        if (isset($this->keepRevisionOf)) {
+            return in_array($relation, $this->keepRevisionOf);
+        }
+
+        if (isset($this->dontKeepRevisionOf)) {
+            return !in_array($relation, $this->dontKeepRevisionOf);
+        }
+
+        return true;
+    }
+
+    /**
+     * Defines the polymorphic relationship
+     *
+     * @return mixed
+     */
+    public function revisionHistory()
+    {
+        return $this->morphMany(Revision::class, 'revisionable');
+    }
+
+    /**
+     * Attempt to find the user id of the currently logged in user.
+     *
+     * @return string|null
+     */
+    private function getUserId()
+    {
+        $jwtAuth = App::make('Tymon\JWTAuth\JWTAuth');
+
+        if ($user = $jwtAuth->user()) {
```

Comment:
@xiphiaz No it does not. Request gets the user by asking the driver registered with AuthManager. So if we make an Auth driver for the jwt implementation, we should be able to get the user from the Request object. That would probably simplify auth a bit overall as well to wrap away the package to a driver where possible, and also prepare us for the new Laravel ACL.